go/vt/vtgate/planbuilder/operators/projection_pushing.go

```diff
@@ -0,0 +1,414 @@
+// LHS expressions to include the derived table. This allows the expressions to be accessed outside
+// the derived table.
+func exposeColumnsThroughDerivedTable(ctx *plancontext.PlanningContext, p *Projection, src *ApplyJoin, lhs, rhs *projector) {
+	derivedTbl, err := ctx.SemTable.TableInfoFor(p.DT.TableID)
+	if err != nil {
+		panic(err)
+	}
+	derivedTblName, err := derivedTbl.Name()
+	if err != nil {
+		panic(err)
+	}
+	lhs.tableName = derivedTblName
+	rhs.tableName = derivedTblName
+
+	lhsIDs := TableID(src.LHS)
+	rhsIDs := TableID(src.RHS)
+	rewriteColumnsForJoin(ctx, src.JoinPredicates.columns, lhsIDs, rhsIDs, lhs, rhs)
+	rewriteColumnsForJoin(ctx, src.JoinColumns.columns, lhsIDs, rhsIDs, lhs, rhs)
+}
+
+func rewriteColumnsForJoin(ctx *plancontext.PlanningContext, columns []applyJoinColumn, lhsIDs, rhsIDs semantics.TableSet, lhs, rhs *projector) {
+	for colIdx, predicate := range columns {
+		for lhsIdx, bve := range predicate.LHSExprs {
+			// since this is on the LHSExprs, we know that dependencies are from that side of the join
+			col := lhs.get(ctx, bve.Expr)
+			predicate.LHSExprs[lhsIdx].Expr = col
+		}
+
+		// now we need to go over the predicate and find
+		var rewriteTo sqlparser.Expr
+
+		pre := func(node, _ sqlparser.SQLNode) bool {
+			expr, ok := node.(sqlparser.Expr)
+			if !ok {
+				return true
+			}
+			deps := ctx.SemTable.RecursiveDeps(expr)
+
+			switch {
+			case deps.IsEmpty():
+				return true
+			case deps.IsSolvedBy(lhsIDs):
+				rewriteTo = lhs.get(ctx, expr)
+				return false
+			case deps.IsSolvedBy(rhsIDs):
+				rewriteTo = rhs.get(ctx, expr)
+				return false
+			default:
+				return true
+			}
+		}
+
+		post := func(cursor *sqlparser.CopyOnWriteCursor) {
+			if rewriteTo != nil {
+				cursor.Replace(rewriteTo)
+				rewriteTo = nil
+				return
+			}
+		}
+		newOriginal := sqlparser.CopyOnRewrite(predicate.Original, pre, post, ctx.SemTable.CopySemanticInfo).(sqlparser.Expr)
+		predicate.Original = newOriginal
+
+		columns[colIdx] = predicate
+	}
+}
+
+// prefixColNames adds qualifier prefixes to all ColName:s.
+// We want to be more explicit than the user was to make sure we never produce invalid SQL
+func prefixColNames(ctx *plancontext.PlanningContext, tblName sqlparser.TableName, e sqlparser.Expr) sqlparser.Expr {
+	return sqlparser.CopyOnRewrite(e, nil, func(cursor *sqlparser.CopyOnWriteCursor) {
+		col, ok := cursor.Node().(*sqlparser.ColName)
+		if !ok {
+			return
+		}
+		cursor.Replace(sqlparser.NewColNameWithQualifier(col.Name.String(), tblName))
+	}, ctx.SemTable.CopySemanticInfo).(sqlparser.Expr)
+}
+
+func createProjectionWithTheseColumns(
+	ctx *plancontext.PlanningContext,
+	src Operator,
+	p *projector,
+	dt *DerivedTable,
+) Operator {
+	if len(p.columns) == 0 {
+		return src
+	}
+	proj := createProjection(ctx, src, "")
+	proj.Columns = AliasedProjections(p.columns)
+	if dt != nil {
+		kopy := *dt
```

Comment:
we do not have a proper clone function for `DerivedTable`?

Reply:
nope. I'll add one

Reply:
hmm... I changed my mind. Don't think we need one. This is the only place we try to clone it, and it's not a pure cloning anyway